### ISSUE TITLE
Correct reporting of oneOf (anyOf, allOf) errors

### ISSUE THREAD

**Issue by jpmckinney** [OpenDataServices/cove#895]
e.g. http://standard.open-contracting.org/validator/data/efa21228-0d98-4080-8c4f-ab63db58e1f6

"[very long JSON] is not valid under any of the given schemas"

The hundreds of rows of `oneOf` errors should be collected into a single error, not presented as distinct errors.

If that's not possible, then if we can collapse the long JSON (e.g. show a snippet and require the user to click on something to see the full JSON), it will make reviewing errors easier.

**Comment by robredpath** [comment]
@duncandewhurst this is the issue you were looking for

**Comment by jpmckinney** [comment]
@robredpath mentioned having a solution from another standard (BODS, I think).

**Comment by jpmckinney** [comment]
@robredpath Was this completed as part of the April sprint? (The sprint doc hadn't been updated to document the outputs.)

**Comment by robredpath** [comment]
@jpmckinney oops, sorry - I've updated the doc now. This wasn't done as part of the April sprint.

**Comment by duncandewhurst** [comment]
I came across [another variant](http://standard.open-contracting.org/review/data/a842e150-06a3-4c8a-b98d-e121919638e3) of this issue today when validating a record package from Uganda, the path to the error was `records`:

> (giant block of JSON) has non-unique elements.

There is also something weird going on because `OrderedDict` is appearing in the big block of JSON, when it doesn't appear in the raw data:

![image](https://user-images.githubusercontent.com/19265814/61698754-881b3d00-ad31-11e9-82a6-83632427a220.png)

**Comment by duncandewhurst** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [Data Review Tool: Correct reporting of giant blocks of JSON](https://trello.com/c/dXzzBjKM/48-data-review-tool-correct-reporting-of-giant-blocks-of-json)

**Comment by Bjwebb** [comment]
> How much work would it be to get individual reports of errors in the embedded releases? I think that would make the most sense for users

Yes, we can do this in a few days. It's what we did for BODS, [see these screenshots for an example](https://github.com/openownership/cove-bods/issues/16#issuecomment-460512901). We replace the `oneOf` validation function from jsonschema [with our own custom one](https://github.com/OpenDataServices/lib-cove/blob/61ed8d096817c52a845ac7b1e84ae5b11e19af77/libcove/lib/common.py#L85), so we can modify this further to handle the OCDS case better.

The reason we have to do this on a case by case basis, rather than improving the generic `oneOf` function, is that in general there's no way of telling which sub-schema you intended to use.

For BODS we can tell which subschema is meant because they each have a different `statementType`.

In the OCDS case we need to tell whether you intended Linked releases or Embedded releases. The easiest way to do this is to look for the `url` field, as we only expect this for Linked releases.

Additionally, for BODS a single statement is nested within the `oneOf`, whereas for OCDS, the whole list of releases is. This means we need to handle the case where some releases look like Linked releases and others like Embedded releases.

> (giant block of JSON) has non-unique elements.

This is a different case, so won't be affected by the above work. It is easy to just remove the large block. I suggest creating a new issue for doing any more than this.

> There is also something weird going on because `OrderedDict` is appearing in the big block of JSON, when it doesn't appear in the raw data:

This is because the jsonschema library displays the Python representation of the data, rather than the JSON representation. (Whereas in our custom code for oneOf we use the JSON representation).

**Comment by robredpath** [comment]
Thanks, @Bjwebb 

> In the OCDS case we need to tell whether you intended Linked releases or Embedded releases. The easiest way to do this is to look for the `url` field, as we only expect this for Linked releases.

Presumably this only works if the error the user has made isn't to do with that field? This certainly sounds sufficient for a dramatic improvement over current behaviour, but am I right in thinking this algorithm could be a good candidate for future improvement to be smarter at detecting what the user was trying to do?

Also - I'd like to document the very helpful rationale here - "we are assuming that this is a Linked Release becuase the url field is only found in Linked Releases" - both in the code and the interface. 

I've created https://github.com/OpenDataServices/cove/issues/1220 for the 'giant block of JSON' issue. 

Presumably when we stop sending large blocks of JSON to the user we'll also stop sending representations of the Python objects that contain them?

**Comment by Bjwebb** [comment]
> Presumably this only works if the error the user has made isn't to do with that field? This certainly sounds sufficient for a dramatic improvement over current behaviour, but am I right in thinking this algorithm could be a good candidate for future improvement to be smarter at detecting what the user was trying to do?

That's right.

Separately, me and @duncandewhurst  discussed that it's better to look for `id`, which only appears in an Embedded release, because there's a proposal to add `url` to the Embedded release also. 

> Also - I'd like to document the very helpful rationale here - "we are assuming that this is a Linked Release becuase the url field is only found in Linked Releases" - both in the code and the interface.

I agree. The interface could be a bit fiddly, but its worth doing.

> Presumably when we stop sending large blocks of JSON to the user we'll also stop sending representations of the Python objects that contain them?

Yes, we'll stop showing the Python representations. Note that they could appear again for different error messages, if we start using certain json schema features (e.g. anyOf instead of oneOf).

**Comment by jpmckinney** [comment]
What specific changes to the results should I be looking for?

> This array should contain either entirely Embedded releases or Linked releases. Embedded releases contain an `id` whereas Linkedreleases do not. Your releases contain a mixture.

Please use sentence case and spaces between 'Linkedreleases'.

**Comment by Bjwebb** [comment]
> What specific changes to the results should I be looking for?

Instead of "valid under any of the given schemas" ([current behaviour](https://standard.open-contracting.org/review/?source_url=https://gist.githubusercontent.com/Bjwebb/f18545c007c31c6655eafbdcc1ff617c/raw/9eae0964be0a3035070d8ce2a21cf1e6a78fd9f4/records_invalid_releases_less_fuzzing.json)) there are now more specific validation messages from one of the subschemas.

**Comment by jpmckinney** [comment]
Aha! It's much better!

**Comment by Bjwebb** [comment]
I've added text about assumptions of linked/embedded releases to the web UI: [[demo]](http://895-oneof-messages.dev.cove.opendataservices.coop/review/?source_url=https://gist.githubusercontent.com/Bjwebb/f18545c007c31c6655eafbdcc1ff617c/raw/aa5a3e06b9baec1b8c5a2e1ffa4c86bdf3a63e66/records_invalid_releases_less_fuzzing_nested.json)

![Screenshot from 2019-10-01 15-17-19](https://user-images.githubusercontent.com/634/65970520-a357a680-e45e-11e9-8fde-a71688efdf2c.png)

Currently these appear for every relevant validation error message, because that was the easiest thing to do. Does this need changing before this work is merged?

**Comment by jpmckinney** [comment]
`date` is required whether it is an embedded or linked release, which produces double the number of error reports. However, only `date` and `tag` have this issue, so I think that's acceptable for now.

Please create an issue to remove this repetition at a later date.

If there are very many errors, the assumptions will be very repetitive. In that case, it might be preferable to group the errors by assumption/subschema. However, this is already an improvement, so can be deployed as-is.

Please create an issue about grouping errors by subschema, so that the text about the assumption is not repeated.

**Comment by Bjwebb** [comment]
New issues:
Remove repeated error reports - https://github.com/OpenDataServices/cove/issues/1234
Grouping errors by subschema - https://github.com/OpenDataServices/cove/issues/1233

**Comment by Bjwebb** [comment]
This was deployed to live couple of weeks ago.
Example link https://standard.open-contracting.org/review/?source_url=https://gist.githubusercontent.com/Bjwebb/f18545c007c31c6655eafbdcc1ff617c/raw/aa5a3e06b9baec1b8c5a2e1ffa4c86bdf3a63e66/records_invalid_releases_less_fuzzing_nested.json

**Comment by duncandewhurst** [comment]
This issue still seems to be present with Zambia's data: [https://standard.open-contracting.org/review/data/d7b728ba-e43b-462e-9cac-6e24e5b12325](https://standard.open-contracting.org/review/data/d7b728ba-e43b-462e-9cac-6e24e5b12325)

To reproduce, download the July 2019 record package from ZPPA's [download page](https://www.zppa.org.zm/record-packages) and submit it to CoVE.

**Comment by Bjwebb** [comment]
Looks as I'd expect. Could you check again and screenshot the problem?

**Comment by duncandewhurst** [comment]
Hmm on clicking the link above it seems to be resolved, but on uploading the file afresh the issue reappears:

![image](https://user-images.githubusercontent.com/19265814/68893192-81558180-0789-11ea-924b-eb8e6e3851c0.png)

Zambia's portal seems to be down at the moment, so I uploaded the file [here](https://drive.google.com/file/d/1MzAOjbXzLiCrtk8LHwi3AtzxspEsazR7/view?usp=sharing)

**Comment by Bjwebb** [comment]
I've identified the source of the problem here, which is my code doesn't work properly with 1.0. This is because it looks for the `title` of the `releases` array, which isn't there in 1.0. I'm guessing the reason it sometimes works is that you've selected 1.1 from the dropdown.

Today I've fixed a bug with arrays being falsely flagged as non-unique https://github.com/OpenDataServices/cove/issues/1246. There's now no validation errors with that Zambia file, including none to do with oneOfs. The underlying problem is still there though, and can be reproduced by changing the version in any of the fixtures.
